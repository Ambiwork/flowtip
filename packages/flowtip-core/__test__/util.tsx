const scale = {x: 1, y: 0.5};

const lines = {
  light: {h: '─', v: '│', tl: '┌', tr: '┐', br: '┘', bl: '└'},
  heavy: {h: '━', v: '┃', tl: '┏', tr: '┓', br: '┛', bl: '┗'},
  dashed: {h: '┄', v: '┆', tl: '┌', tr: '┐', br: '┘', bl: '└'},
};

const tail = {
  top: '┸',
  left: '┥',
  bottom: '┰',
  right: '┝',
};

const draw = (canvas, x, y, c) => {
  if (x >= 0 && y >= 0) canvas[y][x] = c;
};

const drawScaled = (canvas, x, y, c) => {
  const _x = Math.round(x * scale.x);
  const _y = Math.round(y * scale.y);

  if (_x >= 0 && _y >= 0) canvas[_y][_x] = c;
};

const drawRect = (canvas = [], rect, style = 'light') => {
  const _top = Math.round(rect.top * scale.y);
  const _left = Math.round(rect.left * scale.x);
  const _bottom = Math.round(rect.top * scale.y + rect.height * scale.y);
  const _right = Math.round(rect.left * scale.x + rect.width * scale.x);
  const _height = Math.round(rect.height * scale.y);
  const _width = Math.round(rect.width * scale.x);

  for (let y = 0, yLen = _bottom; y <= yLen; y++) {
    canvas[y] = canvas[y] || [];
  }

  if (_height !== 0 && _width !== 0) {
    draw(canvas, _left, _top, lines[style].tl);
    draw(canvas, _right, _top, lines[style].tr);
    draw(canvas, _right, _bottom, lines[style].br);
    draw(canvas, _left, _bottom, lines[style].bl);

    for (let y = _top + 1; y < _bottom; y++) {
      draw(canvas, _left, y, lines[style].v);
      draw(canvas, _right, y, lines[style].v);
    }

    for (let x = _left + 1; x < _right; x++) {
      draw(canvas, x, _top, lines[style].h);
      draw(canvas, x, _bottom, lines[style].h);
    }
  } else if (_height === 0) {
    for (let x = _left; x <= _right; x++) {
      draw(canvas, x, _top, lines[style].h);
    }
  } else {
    for (let y = _top; y <= _bottom; y++) {
      draw(canvas, _left, y, lines[style].v);
    }
  }

  return canvas;
};

const printCanvas = (canvas) => {
  let result = '';

  for (let y = 0, yLen = canvas.length; y < yLen; y++) {
    for (let x = 0, xLen = canvas[y].length; x < xLen; x++) {
      result += canvas[y][x] || ' ';
    }

    result += '\n';
  }

  return result.slice(0, -1);
};

export const translateRect = (rect, left, top) => ({
  left: rect.left - left,
  top: rect.top - top,
  right: rect.right - left,
  bottom: rect.bottom - top,
  width: rect.width,
  height: rect.height,
});

export const drawResult = (config, result) => {
  const {bounds, target} = config;
  const {rect, region, reason, overlapCenter} = result;

  const top = Math.min(rect.top, bounds.top, target.top);
  const left = Math.min(rect.left, bounds.left, target.left);

  const _bounds = translateRect(bounds, left, top);
  const _target = translateRect(target, left, top);
  const _rect = translateRect(rect, left, top);
  const canvas = [];

  drawRect(canvas, _bounds, 'dashed');
  drawRect(canvas, _target, 'heavy');
  drawRect(canvas, _rect, 'light');

  if (region === 'bottom') {
    drawScaled(canvas, _rect.left + overlapCenter, _rect.top, tail.top);
  } else if (region === 'top') {
    drawScaled(canvas, _rect.left + overlapCenter, _rect.bottom, tail.bottom);
  } else if (region === 'left') {
    drawScaled(canvas, _rect.right, _rect.top + overlapCenter, tail.right);
  } else if (region === 'right') {
    drawScaled(canvas, _rect.left, _rect.top + overlapCenter, tail.left);
  }

  const printed = printCanvas(canvas)
    .split('\n')
    .join(`\n${region[0]}  `);

  return `
  ${reason}
  ${region}

${region[0]}  ${printed}
  `;
};
