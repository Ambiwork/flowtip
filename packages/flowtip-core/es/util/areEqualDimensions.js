var areEqualDimensions = function areEqualDimensions(a, b) {
  if (a === b) return true;

  if ((a === null || a === undefined) && (b === null || b === undefined)) {
    return true;
  }

  if (a === null || a === undefined || b === null || b === undefined) {
    return false;
  }

  return a.width === b.width && a.height === b.height;
};

export default areEqualDimensions;
//# sourceMappingURL=areEqualDimensions.js.map