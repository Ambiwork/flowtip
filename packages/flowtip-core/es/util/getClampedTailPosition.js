import { LEFT, RIGHT } from "../flowtip";

var getClampedTailPosition = function getClampedTailPosition(result, tail, tailOffset) {
  var region = result.region,
      rect = result.rect,
      overlapCenter = result.overlapCenter;
  var offset;
  var range;

  if (region === RIGHT || region === LEFT) {
    offset = tail.height / 2;
    range = rect.height;
  } else {
    // Position is top or bottom.
    offset = tail.width / 2;
    range = rect.width;
  }

  var min = offset + tailOffset;
  var max = range - min;
  return Math.min(max, Math.max(min, overlapCenter)) - offset;
};

export default getClampedTailPosition;
//# sourceMappingURL=getClampedTailPosition.js.map