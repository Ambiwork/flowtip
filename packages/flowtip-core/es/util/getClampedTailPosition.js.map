{"version":3,"sources":["../../src/util/getClampedTailPosition.js"],"names":["LEFT","RIGHT","getClampedTailPosition","result","tail","tailOffset","region","rect","overlapCenter","offset","range","height","width","min","max","Math"],"mappings":"AACA,SAAQA,IAAR,EAAcC,KAAd;;AAGA,IAAMC,yBAAyB,SAAzBA,sBAAyB,CAC7BC,MAD6B,EAE7BC,IAF6B,EAG7BC,UAH6B,EAIlB;AAAA,MACJC,MADI,GAC2BH,MAD3B,CACJG,MADI;AAAA,MACIC,IADJ,GAC2BJ,MAD3B,CACII,IADJ;AAAA,MACUC,aADV,GAC2BL,MAD3B,CACUK,aADV;AAEX,MAAIC,MAAJ;AACA,MAAIC,KAAJ;;AAEA,MAAIJ,WAAWL,KAAX,IAAoBK,WAAWN,IAAnC,EAAyC;AACvCS,aAASL,KAAKO,MAAL,GAAc,CAAvB;AACAD,YAAQH,KAAKI,MAAb;AACD,GAHD,MAGO;AACL;AACAF,aAASL,KAAKQ,KAAL,GAAa,CAAtB;AACAF,YAAQH,KAAKK,KAAb;AACD;;AAED,MAAMC,MAAMJ,SAASJ,UAArB;AACA,MAAMS,MAAMJ,QAAQG,GAApB;AAEA,SAAOE,KAAKF,GAAL,CAASC,GAAT,EAAcC,KAAKD,GAAL,CAASD,GAAT,EAAcL,aAAd,CAAd,IAA8CC,MAArD;AACD,CAtBD;;AAwBA,eAAeP,sBAAf","sourcesContent":["// @flow\nimport {LEFT, RIGHT} from '../flowtip';\nimport type {Dimensions, Result} from '../flowtip';\n\nconst getClampedTailPosition = (\n  result: Result,\n  tail: Dimensions,\n  tailOffset: number,\n): number => {\n  const {region, rect, overlapCenter} = result;\n  let offset;\n  let range;\n\n  if (region === RIGHT || region === LEFT) {\n    offset = tail.height / 2;\n    range = rect.height;\n  } else {\n    // Position is top or bottom.\n    offset = tail.width / 2;\n    range = rect.width;\n  }\n\n  const min = offset + tailOffset;\n  const max = range - min;\n\n  return Math.min(max, Math.max(min, overlapCenter)) - offset;\n};\n\nexport default getClampedTailPosition;\n"],"file":"getClampedTailPosition.js"}