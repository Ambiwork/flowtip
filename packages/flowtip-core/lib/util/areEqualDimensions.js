"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var areEqualDimensions = function areEqualDimensions(a, b) {
  if (a === b) return true;

  if ((a === null || a === undefined) && (b === null || b === undefined)) {
    return true;
  }

  if (a === null || a === undefined || b === null || b === undefined) {
    return false;
  }

  return a.width === b.width && a.height === b.height;
};

var _default = areEqualDimensions;
exports.default = _default;
//# sourceMappingURL=areEqualDimensions.js.map