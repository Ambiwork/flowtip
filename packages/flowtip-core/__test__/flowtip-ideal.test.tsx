import flowtip from '../src/flowtip';

import {drawResult} from './util';

import {
  internalTop,
  externalTop,
  internalBottom,
  externalBottom,
  internalLeft,
  externalLeft,
  internalRight,
  externalRight,
  oversizedHorizontal,
  oversizedVertical,
} from './fixtures';

describe('flowtip', () => {
  describe('ideal result', () => {
    describe('at internal top', () => {
      it('should return `bottom`', () => {
        const config = internalTop;
        const result = flowtip(config);

        expect(drawResult(config, result)).toMatchSnapshot();
        expect(result.reason).toEqual('ideal');
        expect(result.region).toEqual('bottom');
      });
    });

    describe('at external top', () => {
      it('should return `bottom`', () => {
        const config = externalTop;
        const result = flowtip(config);

        expect(drawResult(config, result)).toMatchSnapshot();
        expect(result.reason).toEqual('ideal');
        expect(result.region).toEqual('bottom');
      });
    });

    describe('at internal bottom', () => {
      it('should return `top`', () => {
        const config = internalBottom;
        const result = flowtip(config);

        expect(drawResult(config, result)).toMatchSnapshot();
        expect(result.reason).toEqual('ideal');
        expect(result.region).toEqual('top');
      });
    });

    describe('at external bottom', () => {
      it('should return `top`', () => {
        const config = externalBottom;
        const result = flowtip(config);

        expect(drawResult(config, result)).toMatchSnapshot();
        expect(result.reason).toEqual('ideal');
        expect(result.region).toEqual('top');
      });
    });

    describe('at internal left', () => {
      it('should return `right`', () => {
        const config = internalLeft;
        const result = flowtip(config);

        expect(drawResult(config, result)).toMatchSnapshot();
        expect(result.reason).toEqual('ideal');
        expect(result.region).toEqual('right');
      });
    });

    describe('at external left', () => {
      it('should return `right`', () => {
        const config = externalLeft;
        const result = flowtip(config);

        expect(drawResult(config, result)).toMatchSnapshot();
        expect(result.reason).toEqual('ideal');
        expect(result.region).toEqual('right');
      });
    });

    describe('at internal right', () => {
      it('should return `left`', () => {
        const config = internalRight;
        const result = flowtip(config);

        expect(drawResult(config, result)).toMatchSnapshot();
        expect(result.reason).toEqual('ideal');
        expect(result.region).toEqual('left');
      });
    });

    describe('at external right', () => {
      it('should return `left`', () => {
        const config = externalRight;
        const result = flowtip(config);

        expect(drawResult(config, result)).toMatchSnapshot();
        expect(result.reason).toEqual('ideal');
        expect(result.region).toEqual('left');
      });
    });

    describe('with oversized content', () => {
      it('should center horizontally', () => {
        const config = oversizedHorizontal;
        const result = flowtip(config);

        expect(drawResult(config, result)).toMatchSnapshot();
        expect(result.reason).toEqual('ideal');
        expect(result.region).toEqual('bottom');
      });

      it('should center vertically', () => {
        const config = oversizedVertical;
        const result = flowtip(config);

        expect(drawResult(config, result)).toMatchSnapshot();
        expect(result.reason).toEqual('ideal');
        expect(result.region).toEqual('right');
      });
    });

    describe('with region config', () => {
      it('should override `default` reason', () => {
        const config = {
          ...internalTop,
          region: 'bottom',
        };
        const result = flowtip(config);

        expect(drawResult(config, result)).toMatchSnapshot();
        expect(result.reason).toEqual('ideal');
        expect(result.region).toEqual('bottom');
      });

      it('should return `inverted` reason', () => {
        const config = {
          ...internalTop,
          region: 'top',
          disabled: {top: true},
        };
        const result = flowtip(config);

        expect(drawResult(config, result)).toMatchSnapshot();
        expect(result.reason).toEqual('ideal');
        expect(result.region).toEqual('bottom');
      });
    });
  });
});
