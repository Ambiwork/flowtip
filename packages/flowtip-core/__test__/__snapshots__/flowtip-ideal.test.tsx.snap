// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flowtip ideal result at external bottom should return \`top\` 1`] = `
"
  ideal
  top

t  ┌┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┐
t  ┆                                       ┆
t  ┆                                       ┆
t  ┆                                       ┆
t  ┆                                       ┆
t  ┆                                       ┆
t  ┆                                       ┆
t  ┆                                       ┆
t  ┆                                       ┆
t  ┆                                       ┆
t  ┆                                       ┆
t  ┆                                       ┆
t  ┆                                       ┆
t  ┆                                       ┆
t  ┆                                       ┆
t  ┆                                       ┆
t  ┆             ┌─────────┐               ┆
t  ┆             │         │               ┆
t  ┆             └────┰────┘               ┆
t  ┆                                       ┆
t  └┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┘
t  
t  
t                   ┏━━━┓
t                   ┃   ┃
t                   ┗━━━┛
  "
`;

exports[`flowtip ideal result at external left should return \`right\` 1`] = `
"
  ideal
  right

r                 ┌┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┐
r                 ┆                                       ┆
r                 ┆                                       ┆
r                 ┆                                       ┆
r                 ┆                                       ┆
r                 ┆                                       ┆
r                 ┆                                       ┆
r                 ┆                                       ┆
r                 ┆                                       ┆
r                 ┆                                       ┆
r  ┏━━━┓          ┆   ┌─────────┐                         ┆
r  ┃   ┃          ┆   ┥         │                         ┆
r  ┗━━━┛          ┆   └─────────┘                         ┆
r                 ┆                                       ┆
r                 ┆                                       ┆
r                 ┆                                       ┆
r                 ┆                                       ┆
r                 ┆                                       ┆
r                 ┆                                       ┆
r                 ┆                                       ┆
r                 └┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┘
  "
`;

exports[`flowtip ideal result at external right should return \`left\` 1`] = `
"
  ideal
  left

l  ┌┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┐
l  ┆                                       ┆
l  ┆                                       ┆
l  ┆                                       ┆
l  ┆                                       ┆
l  ┆                                       ┆
l  ┆                                       ┆
l  ┆                                       ┆
l  ┆                                       ┆
l  ┆                                       ┆
l  ┆                         ┌─────────┐   ┆    ┏━━━┓
l  ┆                         │         ┝   ┆    ┃   ┃
l  ┆                         └─────────┘   ┆    ┗━━━┛
l  ┆                                       ┆
l  ┆                                       ┆
l  ┆                                       ┆
l  ┆                                       ┆
l  ┆                                       ┆
l  ┆                                       ┆
l  ┆                                       ┆
l  └┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┘
  "
`;

exports[`flowtip ideal result at external top should return \`bottom\` 1`] = `
"
  ideal
  bottom

b                   ┏━━━┓
b                   ┃   ┃
b                   ┗━━━┛
b  
b  
b  ┌┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┐
b  ┆                                       ┆
b  ┆             ┌────┸────┐               ┆
b  ┆             │         │               ┆
b  ┆             │         │               ┆
b  ┆             └─────────┘               ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  └┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┘
  "
`;

exports[`flowtip ideal result at internal bottom should return \`top\` 1`] = `
"
  ideal
  top

t  ┌┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┐
t  ┆                                       ┆
t  ┆                                       ┆
t  ┆                                       ┆
t  ┆                                       ┆
t  ┆                                       ┆
t  ┆                                       ┆
t  ┆                                       ┆
t  ┆             ┌─────────┐               ┆
t  ┆             │         │               ┆
t  ┆             │         │               ┆
t  ┆             └────┰────┘               ┆
t  ┆                                       ┆
t  ┆                ┏━━━┓                  ┆
t  ┆                ┃   ┃                  ┆
t  ┆                ┗━━━┛                  ┆
t  ┆                                       ┆
t  ┆                                       ┆
t  ┆                                       ┆
t  ┆                                       ┆
t  └┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┘
  "
`;

exports[`flowtip ideal result at internal left should return \`right\` 1`] = `
"
  ideal
  right

r  ┌┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┐
r  ┆                                       ┆
r  ┆                                       ┆
r  ┆                                       ┆
r  ┆                                       ┆
r  ┆                                       ┆
r  ┆                                       ┆
r  ┆                                       ┆
r  ┆                                       ┆
r  ┆                                       ┆
r  ┆    ┏━━━┓   ┌─────────┐                ┆
r  ┆    ┃   ┃   ┥         │                ┆
r  ┆    ┗━━━┛   └─────────┘                ┆
r  ┆                                       ┆
r  ┆                                       ┆
r  ┆                                       ┆
r  ┆                                       ┆
r  ┆                                       ┆
r  ┆                                       ┆
r  ┆                                       ┆
r  └┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┘
  "
`;

exports[`flowtip ideal result at internal right should return \`left\` 1`] = `
"
  ideal
  left

l  ┌┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┐
l  ┆                                       ┆
l  ┆                                       ┆
l  ┆                                       ┆
l  ┆                                       ┆
l  ┆                                       ┆
l  ┆                                       ┆
l  ┆                                       ┆
l  ┆                                       ┆
l  ┆                                       ┆
l  ┆               ┌─────────┐   ┏━━━┓     ┆
l  ┆               │         ┝   ┃   ┃     ┆
l  ┆               └─────────┘   ┗━━━┛     ┆
l  ┆                                       ┆
l  ┆                                       ┆
l  ┆                                       ┆
l  ┆                                       ┆
l  ┆                                       ┆
l  ┆                                       ┆
l  ┆                                       ┆
l  └┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┘
  "
`;

exports[`flowtip ideal result at internal top should return \`bottom\` 1`] = `
"
  ideal
  bottom

b  ┌┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┐
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                ┏━━━┓                  ┆
b  ┆                ┃   ┃                  ┆
b  ┆                ┗━━━┛                  ┆
b  ┆                                       ┆
b  ┆             ┌────┸────┐               ┆
b  ┆             │         │               ┆
b  ┆             │         │               ┆
b  ┆             └─────────┘               ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  └┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┘
  "
`;

exports[`flowtip ideal result with oversized content should center horizontally 1`] = `
"
  ideal
  bottom

b            ┌┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┐
b            ┆                                       ┆
b            ┆   ┏━━━┓                               ┆
b            ┆   ┃   ┃                               ┆
b            ┆   ┗━━━┛                               ┆
b            ┆                                       ┆
b  ┌───────────────┸───────────────────────────────────────────┐
b  │         ┆                                       ┆         │
b  │         ┆                                       ┆         │
b  └───────────────────────────────────────────────────────────┘
b            ┆                                       ┆
b            ┆                                       ┆
b            ┆                                       ┆
b            ┆                                       ┆
b            ┆                                       ┆
b            ┆                                       ┆
b            ┆                                       ┆
b            ┆                                       ┆
b            ┆                                       ┆
b            ┆                                       ┆
b            └┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┘
  "
`;

exports[`flowtip ideal result with oversized content should center vertically 1`] = `
"
  ideal
  right

r              ┌─────────┐
r              │         │
r              │         │
r              │         │
r              │         │
r  ┌┄┄┄┄┄┄┄┄┄┄┄│┄┄┄┄┄┄┄┄┄│┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┐
r  ┆           │         │                 ┆
r  ┆   ┏━━━┓   │         │                 ┆
r  ┆   ┃   ┃   ┥         │                 ┆
r  ┆   ┗━━━┛   │         │                 ┆
r  ┆           │         │                 ┆
r  ┆           │         │                 ┆
r  ┆           │         │                 ┆
r  ┆           │         │                 ┆
r  ┆           │         │                 ┆
r  ┆           │         │                 ┆
r  ┆           │         │                 ┆
r  ┆           │         │                 ┆
r  ┆           │         │                 ┆
r  ┆           │         │                 ┆
r  ┆           │         │                 ┆
r  ┆           │         │                 ┆
r  ┆           │         │                 ┆
r  ┆           │         │                 ┆
r  ┆           │         │                 ┆
r  └┄┄┄┄┄┄┄┄┄┄┄│┄┄┄┄┄┄┄┄┄│┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┘
r              │         │
r              │         │
r              │         │
r              │         │
r              └─────────┘
  "
`;

exports[`flowtip ideal result with region config should override \`default\` reason 1`] = `
"
  ideal
  bottom

b  ┌┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┐
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                ┏━━━┓                  ┆
b  ┆                ┃   ┃                  ┆
b  ┆                ┗━━━┛                  ┆
b  ┆                                       ┆
b  ┆             ┌────┸────┐               ┆
b  ┆             │         │               ┆
b  ┆             │         │               ┆
b  ┆             └─────────┘               ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  └┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┘
  "
`;

exports[`flowtip ideal result with region config should return \`inverted\` reason 1`] = `
"
  ideal
  bottom

b  ┌┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┐
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                ┏━━━┓                  ┆
b  ┆                ┃   ┃                  ┆
b  ┆                ┗━━━┛                  ┆
b  ┆                                       ┆
b  ┆             ┌────┸────┐               ┆
b  ┆             │         │               ┆
b  ┆             │         │               ┆
b  ┆             └─────────┘               ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  ┆                                       ┆
b  └┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┘
  "
`;
